[5991 + 5987] add model and form mixin for users's agreement on orgsnisation's terms of use

fuzzylogic2000 opened the pull request:
I put it all into the mixin instead of using rules.

fuzzylogic2000 commented on apps/organisations/mixins.py:
```diff
@@ -0,0 +1,48 @@
+from apps.organisations.models import OrganisationTermsOfUse
+
+ORGANISATION_TERMS_OF_USE_LABEL = _(
+    'Yes, I have read and agree to the terms of use. '
```
The label probably has to go to the __init__, so that we can actually add the link to the organisation's terms of use.

phillimorland commented on apps/contrib/templates/a4_candy_contrib/includes/form_checkbox_field.html:
```diff
@@ -3,7 +3,7 @@
     <label class="form-check__label">
         {{ field }}
-        {{ field.label }}
+        {{ field.label }}{% if field.field.required %}<span role="presentation" title="{% trans 'This field is required' %}">*</span>{% endif %}
```
no clue but only place I know we have a required checkbox, (sign up form) we don't use it?!

phillimorland replied:
I think it's from mB so we could check there?

Rineee replied:
@fuzzylogic2000 Was it maybe, because for the image_right_of_use, it is only required if there is an actual image (and that is checked in the clean)? So maybe for organisations_terms_of_use the form field could be required?

Rineee replied:
Haha, ok, maybe should have read this conversation til the end, sorryyy :see_no_evil:

phillimorland reviewed (commented):
Code looks good but will wait for Rine for proper code review.
When testing and attempt to save without checking box it only has the browser error handling, is this intended? also maybe this is why we didn't have required on the checkbox as when we have error handling this would mean there would double errors?

fuzzylogic2000 commented:
@phillimorland I changed the error handling and added the link to the label. Adding the second sentence as a helptext (not quite the design) is confirmed by @mcastro-lqd .
Sooo, we could merge to try it out. But then this would get half-finished into kosmo. :(

phillimorland reviewed (approved):
fabulous error handling :)

Rineee commented on apps/organisations/mixins.py:
```diff
@@ -0,0 +1,76 @@
+    def save(self, commit=True):
+        instance = super().save(commit=commit)
+        if commit and 'organisation_terms_of_use' in self.cleaned_data:
+            try:
+                agreed = OrganisationTermsOfUse.objects.get(
+                    user=self.user,
+                    organisation=self.module.project.organisation
+                )
+                agreed.has_agreed = \
+                    self.cleaned_data['organisation_terms_of_use']
+                agreed.save()
+            except OrganisationTermsOfUse.DoesNotExist:
+                OrganisationTermsOfUse.objects.create(
+                    user=self.user,
+                    organisation=self.module.project.organisation,
+                    has_agreed=self.cleaned_data['organisation_terms_of_use']
+                )
```
Maybe this could be handled with get_or_create()?

Rineee replied:
Ah ok, didn't know that! I thought that this is exactly what get_or_create() does, so try: get, except: create. But if this is more robust, lets leave it like that.

Rineee replied:
Should we then also remove get_or_create in other places, you think?

Rineee commented on apps/organisations/mixins.py:
```diff
@@ -0,0 +1,76 @@
+        self.fields['organisation_terms_of_use'].label = \
+            self._get_terms_of_use_label
+
+        if OrganisationTermsOfUse.objects.filter(
+            user=self.user,
+            organisation=self.module.project.organisation
+        ).exists() and OrganisationTermsOfUse.objects.get(
+            user=self.user,
+            organisation=self.module.project.organisation
+        ).has_agreed is True:
```
Wouldn't it be the same to just check if the entry with has_agreed=True exists? Then it would only be one query.. So I mean
`if OrganisationTermsOfUse.objects.filter(
            user=self.user,
            organisation=self.module.project.organisation,
            has_agreed=True
        ).exists()`

Rineee replied:
And maybe it could go into the model as a property or method? Sth like user_has_agreed(user, organisation)? But not sure if that makes sense.. just thought we will probably need this again at other places.

fuzzylogic2000 replied:
Yes, both good ideas! Which one should I do now? :)
Will start with trying the second version.

fuzzylogic2000 replied:
I put it into the user model, because that is where we also should put the list for the overview.

Rineee replied:
Perfect!

Rineee commented:
@fuzzylogic2000 Supercool, it works great! Just made some proposals code wise, but I think this is the way to go!

fuzzylogic2000 commented on apps/ideas/views.py:
```diff
@@ -109,6 +109,11 @@ class IdeaCreateView(AbstractIdeaCreateView):
+        kwargs = super().get_form_kwargs()
+        kwargs['user'] = self.request.user
+        return kwargs
+
```
Yes, it was there first, but then all the tests are failing because the mixin is not yet there. We can either add it to all forms right away or keep it like that, test in idea on dev and then move it back when we add it to all the other modules.

Rineee replied:
Sure, we can also move it there later, didn't think about failing tests!

Rineee commented on apps/users/models.py:
```diff
@@ -148,3 +149,10 @@ def get_full_name(self):
+            user=self,
+            organisation=organisation,
+            has_agreed=True
+        ).exists()
```
Fabulous!

Rineee commented on apps/organisations/mixins.py:
```diff
@@ -0,0 +1,65 @@
+
+        if self.user.has_agreed_on_org_terms(
+                self.module.project.organisation):
+            del self.fields['organisation_terms_of_use']
```
So clever to just delete the field then!

Rineee reviewed (approved):
Yeah, super nice!
